tests/semantic_analyser.cpp

```diff
@@ -8,237 +11,150 @@
-  assert(
-      expected_result != 0 &&
-      "Use test(MockBPFfeature&, const std::string&) for expected successes");
-  auto bpftrace = get_mock_bpftrace();
-  bool mock_has_features = feature.has_features_;
-  return test(*bpftrace,
-              mock_has_features,
-              input,
-              expected_result,
-              {},
-              safe_mode,
-              false);
-}
-
-ast::ASTContext test(const std::string &input,
-                     int expected_result,
-                     bool safe_mode,
-                     bool has_child = false)
-{
-  auto bpftrace = get_mock_bpftrace();
-  return test(
-      *bpftrace, true, input, expected_result, {}, safe_mode, has_child);
-}
-
-ast::ASTContext test(const std::string &input, int expected_result)
-{
-  // This function will eventually be deprecated in favour of test_error()
-  assert(expected_result != 0 &&
-         "Use test(const std::string&) for expected successes");
-  auto bpftrace = get_mock_bpftrace();
-  return test(*bpftrace, true, input, expected_result, {}, true, false);
-}
+// This exists as a test fixture because the types may refer to `bpftrace`, so
+// this objects lifetime must exceed the tests lifetime. This is easier with a
+// fixture, and allows us to have a single harness.
+class SemanticAnalyserHarness {
+public:
+  template <Options options = Options{}, typename... Ts>
+    requires((std::is_same_v<std::decay_t<Ts>, Mock> ||
+              std::is_same_v<std::decay_t<Ts>, Warning> ||
+              std::is_same_v<std::decay_t<Ts>, NoWarning> ||
+              std::is_same_v<std::decay_t<Ts>, Error> ||
+              std::is_same_v<std::decay_t<Ts>, ExpectedAST>) &&
+             ...)
+  ast::ASTContext test(std::string_view input, Ts &&...args)
+  {
+    ast::ASTContext ast("stdin", std::string(clean_prefix(input)));
 
-ast::ASTContext test(const std::string &input)
-{
-  auto bpftrace = get_mock_bpftrace();
-  return test(*bpftrace, true, input, 0, {}, true, false);
-}
+    // Reset for each iteration. We only guarantee that the types remain
+    // valid after the ASTContext has been returned.
+    bpftrace_ = get_mock_bpftrace();
 
-ast::ASTContext test(BPFtrace &bpftrace,
-                     const std::string &input,
-                     std::string_view expected_ast)
-{
-  auto ast = test(bpftrace, true, input, 0, {}, true, false);
+    // Extract all extra arguments.
+    auto mock = extract<Mock>(args...);
+    auto warning = extract<Warning>(args...);
+    auto nowarning = extract<NoWarning>(args...);
+    auto error = extract<Error>(args...);
+    auto expected_ast = extract<ExpectedAST>(args...);
 
-  if (expected_ast[0] == '\n')
-    expected_ast.remove_prefix(1); // Remove initial '\n'
+    if (!mock) {
+      mock.emplace(*bpftrace_);
+    }
+    mock->bpftrace.safe_mode_ = options.safe_mode;
+    mock->bpftrace.feature_ = std::make_unique<MockBPFfeature>(
+        options.features);
+    if constexpr (options.with_child) {
```

Comment:
neat - I should do this more

Reply:
It's gone now :) I've gone with your suggestion and will just be using a single form of metaprogramming, which makes it a bit simpler.